bigtable/metrics.go

```diff
@@ -0,0 +1,401 @@
+			},
+			recordedPerAttempt: true,
+		},
+		metricNameServerLatencies: {
+			additionalAttrs: []string{
+				metricLabelKeyStatus,
+				metricLabelKeyStreamingOperation,
+			},
+			recordedPerAttempt: true,
+		},
+		metricNameRetryCount: {
+			additionalAttrs: []string{
+				metricLabelKeyStatus,
+			},
+			recordedPerAttempt: true,
+		},
+	}
+
+	// Generates unique client ID in the format go-<random UUID>@<>hostname
+	generateClientUID = func() (string, error) {
+		hostname := "localhost"
+		hostname, err := os.Hostname()
+		if err != nil {
+			return "", err
+		}
+		return "go-" + uuid.NewString() + "@" + hostname, nil
+	}
+
+	exporterOpts = []option.ClientOption{}
+)
+
+type metricInfo struct {
+	additionalAttrs    []string
+	recordedPerAttempt bool
+}
+
+type builtinMetricsTracerFactory struct {
+	enabled bool
+
+	// To be called on client close
+	shutdown func()
+
+	// attributes that are specific to a client instance and
+	// do not change across different function calls on client
+	clientAttributes []attribute.KeyValue
+
+	operationLatencies metric.Float64Histogram
+	serverLatencies    metric.Float64Histogram
+	attemptLatencies   metric.Float64Histogram
+	retryCount         metric.Int64Counter
+}
+
+func newBuiltinMetricsTracerFactory(ctx context.Context, project, instance, appProfile string, metricsProvider MetricsProvider) (*builtinMetricsTracerFactory, error) {
+	clientUID, err := generateClientUID()
+	if err != nil {
+		log.Printf("built-in metrics: generateClientUID failed: %v. Using empty string in the %v metric atteribute", err, metricLabelKeyClientUID)
+	}
+
+	tracerFactory := &builtinMetricsTracerFactory{
+		enabled: false,
+		clientAttributes: []attribute.KeyValue{
+			attribute.String(monitoredResLabelKeyProject, project),
+			attribute.String(monitoredResLabelKeyInstance, instance),
+			attribute.String(metricLabelKeyAppProfile, appProfile),
+			attribute.String(metricLabelKeyClientUID, clientUID),
+			attribute.String(metricLabelKeyClientName, clientName),
+		},
+		shutdown: func() {},
+	}
+
+	var meterProvider *sdkmetric.MeterProvider
+	if metricsProvider == nil {
+		// Create default meter provider
+		mpOptions, err := builtInMeterProviderOptions(project)
+		if err != nil {
+			return tracerFactory, err
+		}
+		meterProvider = sdkmetric.NewMeterProvider(mpOptions...)
+
+		tracerFactory.enabled = true
+		tracerFactory.shutdown = func() { meterProvider.Shutdown(ctx) }
+	} else {
+		switch metricsProvider.(type) {
+		case NoopMetricsProvider:
+			tracerFactory.enabled = false
+			return tracerFactory, nil
+		default:
+			tracerFactory.enabled = false
+			return tracerFactory, errors.New("unknown MetricsProvider type")
+		}
+	}
+
+	// Create meter and instruments
+	meter := meterProvider.Meter(builtInMetricsMeterName, metric.WithInstrumentationVersion(internal.Version))
+	err = tracerFactory.createInstruments(meter)
+	return tracerFactory, err
+}
+
+func builtInMeterProviderOptions(project string) ([]sdkmetric.Option, error) {
```

Comment:
Could you rename this function and change the `builtIn` part to something more specific to what it does?

Reply:
Changed `MeterProviderOptions` in https://github.com/googleapis/google-cloud-go/pull/10629